OpenTrons/opentrons-api
# refactor(engine): ensure engine/runner shut down gracefully after stop

## Overview

When a `stop` action was issued to a run resource, two bad things would happen if the protocol file portion of the run had not yet been started:

- The hardware controller would be put into an unrecoverable (for the client) state
- The `ProtocolRunner`'s background task would never complete

This PR ensures resolves both those issues.

## Changelog

- Ensure a `ProtocolEngine.halt` recovers the `HardwareAPI`
- Ensure a `ProtocolRunner.stop` cleans up its internal `TaskQueue`
- Rewrite tests for actions router to remove FastAPI from unit tests and remove `verify_response` helper (closes #8394)

## Review requests

Hardware cleanup:

1. Create a basic run
2. Issue a `stop` action
3. Issue a `POST /robot/home`
    - [ ] On `edge`, home errors out with a `ExecutionCancelledError`
    - [ ] On this branch, home should succeed

Task cleanup (easier to test on dev server)

1. Launch dev server
2. Create a basic run
3. Issue a `stop` action
4. Attempt to ctrl-c the server
    - [ ] On `edge`, you'll see `Waiting for background tasks to complete`
    - [ ] On this branch, server will exit

## Risk assessment

Low
